source/MaterialXShaderGen/ShaderGenerators/Glsl/LightGlsl.cpp

```diff
@@ -0,0 +1,112 @@
+#include <MaterialXShaderGen/ShaderGenerators/Glsl/LightGlsl.h>
+#include <MaterialXShaderGen/Util.h>
+
+namespace MaterialX
+{
+
+SgImplementationPtr LightGlsl::creator()
+{
+    return std::make_shared<LightGlsl>();
+}
+
+void LightGlsl::initialize(ElementPtr implementation, ShaderGenerator& shadergen)
+{
+    GlslImplementation::initialize(implementation, shadergen);
+
+    ImplementationPtr impl = implementation->asA<Implementation>();
+    if (!impl)
+    {
+        throw ExceptionShaderGenError("Element '" + implementation->getName() + "' is not an Implementation element");
+    }
+
+    const string& file = impl->getAttribute("file");
+    if (file.empty())
+    {
+        throw ExceptionShaderGenError("No source file specified for implementation '" + impl->getName() + "'");
+    }
+
+    // Find the function name to use
+    _functionName = impl->getAttribute("function");
+    if (_functionName.empty())
+    {
+        // No function given so use nodedef name
+        _functionName = impl->getNodeDefString();
+    }
+
+    _functionSource = "";
+    if (!readFile(shadergen.findSourceCode(file), _functionSource))
+    {
+        throw ExceptionShaderGenError("Can't find source file '" + file + "' used by implementation '" + impl->getName() + "'");
+    }
+
+    NodeDefPtr nodeDef = impl->getNodeDef();
+    _lightUniforms.resize(nodeDef->getInputCount() + nodeDef->getParameterCount());
+
+    size_t index = 0;
+    for (InputPtr input : nodeDef->getInputs())
+    {
+        _lightUniforms[index++] = Shader::Variable(input->getType(), input->getName());
+    }
+    for (ParameterPtr param : nodeDef->getParameters())
+    {
+        _lightUniforms[index++] = Shader::Variable(param->getType(), param->getName());
+    }
+}
+
+void LightGlsl::createVariables(const SgNode& /*node*/, ShaderGenerator& /*shadergen*/, Shader& shader_)
+{
+    HwShader& shader = static_cast<HwShader&>(shader_);
+
+    // Create variables used by this shader
+    for (const Shader::Variable& uniform : _lightUniforms)
+    {
+        shader.createUniform(HwShader::PIXEL_STAGE, HwShader::LIGHT_DATA_BLOCK, uniform.type, uniform.name);
+    }
+
+    // Create uniform for number of active light sources
+    shader.createUniform(HwShader::PIXEL_STAGE, HwShader::PRIVATE_UNIFORMS, DataType::INTEGER, "u_numActiveLightSources",
+        EMPTY_STRING, Value::createValue<int>(1));
```

Comment:
I think I saw this initialization earlier in the review as well. Seems like this should default to 0 instead of 1 as initially there should be no lights set in the generator. If this is set to 1 then it could read through garbage if the app didn't bind any data.

Reply:
Good point, will fix.